nengo_dl/tests/test_converter.py

```diff
@@ -0,0 +1,460 @@
+    #
+    # conv.verify(model, net, training=False, inputs=inp_vals)
+    # conv.verify(model, net, training=True, inputs=inp_vals)
+
+    # test actually converting to nengo objects
+    conv = converter.Converter(model, allow_fallback=False, freeze_batchnorm=True)
+
+    conv.verify(training=False, inputs=inp_vals)
+    with pytest.raises(ValueError, match="does not match output of converted"):
+        # we don't expect the verification to pass for training=True, since we froze
+        # the batch normalization in the nengo network (but not the keras model)
+        conv.verify(training=True, inputs=inp_vals)
+
+    # error if freeze_batchnorm=False
+    with pytest.raises(TypeError, match="unless freeze_batchnorm=True"):
+        converter.Converter(model, allow_fallback=False, freeze_batchnorm=False)
+
+
+def test_relu():
+    inp = tf.keras.Input(shape=(10,))
+    out = tf.keras.layers.ReLU()(inp)
+
+    _test_convert(inp, out, inp_vals=[np.linspace(-1, 1, 10)[None, :]])
+
+
+def test_avg_pool(rng):
+    n_filters = 6
+    inp = tf.keras.Input(shape=(64, n_filters))
+    outputs = [tf.keras.layers.AvgPool1D()(inp)]
+
+    x = tf.keras.layers.Reshape((8, 8, n_filters))(inp)
+    outputs.append(tf.keras.layers.AvgPool2D()(x))
+
+    x = tf.keras.layers.Reshape((8, 8, n_filters))(inp)
+    outputs.append(tf.keras.layers.AvgPool2D(pool_size=3, strides=2)(x))
+
+    x = tf.keras.layers.Reshape((4, 4, 4, n_filters))(inp)
+    outputs.append(tf.keras.layers.AvgPool3D()(x))
+
+    if utils.tf_gpu_installed:
+        x = tf.keras.layers.Reshape((n_filters, 8, 8))(inp)
+        outputs.append(
+            tf.keras.layers.AvgPool2D(
+                pool_size=3, strides=2, data_format="channels_first"
+            )(x)
+        )
+
+    _test_convert(inp, outputs, inp_vals=[rng.uniform(size=(3, 64, n_filters))])
+
+
+def test_global_avg_pool(rng):
+    n_filters = 6
+    inp = tf.keras.Input(shape=(64, n_filters))
+    out0 = tf.keras.layers.GlobalAvgPool1D()(inp)
+
+    x = tf.keras.layers.Reshape((8, 8, n_filters))(inp)
+    out1 = tf.keras.layers.GlobalAvgPool2D()(x)
+
+    x = tf.keras.layers.Reshape((n_filters, 8, 8))(inp)
+    out2 = tf.keras.layers.GlobalAvgPool2D(data_format="channels_first")(x)
+
+    x = tf.keras.layers.Reshape((4, 4, 4, n_filters))(inp)
+    out3 = tf.keras.layers.GlobalAvgPool3D()(x)
+
+    _test_convert(
+        inp, [out0, out1, out2, out3], inp_vals=[rng.uniform(size=(3, 64, n_filters))]
+    )
+
+
+def test_densenet(seed):
+    tf.random.set_seed(seed)
+    model = tf.keras.applications.densenet.DenseNet121(
+        weights=None, include_top=False, input_shape=(224, 224, 3)
+    )
+
+    converter.Converter(
+        model, allow_fallback=False, max_to_avg_pool=True, freeze_batchnorm=True
+    )
+
+    # note: just checking that the conversion passes (i.e. the network is fully
+    # converted); we don't expect any of the verification checks to pass, due to the
+    # max_to_avg_pool swap
```

Comment:
Since we can't call `verify`, it would be nice to be able to assert something here... is there something in the inputs / outputs / net that we can rely on? Maybe even the number of Nengo objects created? I'm guessing you want to avoid running the network on both sims and comparing them.